packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -106,87 +163,196 @@ export class BlobManager {
         return `/${BlobManager.basePath}/${blobId}`;
     }
 
-    public async getBlob(blobId: string): Promise<IFluidHandle<ArrayBufferLike>> {
-        const storageId = this.redirectTable?.get(blobId) ?? blobId;
-        assert(this.hasBlob(storageId), 0x11f /* "requesting unknown blobs" */);
+    private get storageIds(): Set<string> {
+        const ids = new Set<string | undefined>(this.redirectTable.values());
+
+        // If we are detached, we will not have storage IDs, only undefined
+        const havePendingBlobs = ids.delete(undefined);
+
+        // We should not have undefined entries in the table unless we are detached
+        assert(!havePendingBlobs || this.runtime.attachState === AttachState.Detached && ids.size === 0,
+            "redirect table must not have an undefined value while attached");
+
+        return ids as Set<string>;
+    }
+
+    public async getBlob(blobId: string): Promise<ArrayBufferLike> {
+        const pending = this.pendingBlobs.get(blobId);
+        if (pending) {
+            return pending.blob;
+        }
+        let storageId;
+        if (this.runtime.attachState === AttachState.Detached) {
+            assert(this.redirectTable.has(blobId), "requesting unknown blobs");
+            storageId = blobId;
```

Comment:
A comment explaining why we can substitute the blobId for the storageId in this case would be helpful.

I'm guessing that the IDocumentStorageService.readBlob() implementation can resolve blob ids IFF they were created while detached.

This raises the question of why pending blobs are handled inside the BlobManager, while the detached blobs implementation leaks into IDocumentStorageService.  (I'm guessing it has to do with how we store detached vs. pending blobs?)

Reply:
> I'm guessing that the IDocumentStorageService.readBlob() implementation can resolve blob ids IFF they were created while detached.

Yes.
> This raises the question of why pending blobs are handled inside the BlobManager, while the detached blobs implementation leaks into IDocumentStorageService. (I'm guessing it has to do with how we store detached vs. pending blobs?)

This is a good question. While detached, Container stores blobs in an IDetachedBlobStorage. I think we could change this design to treat blobs while detached more similarly to offline, but there are still differences, like the requirement that blobs added while detached must be uploaded before posting the first summary, rather than before emitting "connected" event locally.

Reply:
Thx.  I added a comment about 'storageId = blobId'.